### PR TITLE
bootstrap process will not overwrite existing ssh keys

### PR DESCRIPTION
**Pull request by liangxin1300**
## Motivation
crmsh bootstrap has a very old default of overwriting the root id_rsa ssh key if the customer does not chose the opposite. Also, combined with automation options like -y (that answers yes for all the questions) makes a dangerous combination that can remove SSH access configured previously by the customer, and in extreme cases, could lock the machine depending of the configurations.
#### crm cluster init node

- Generate a new rsa key only if /root/.ssh/id_rsa not exist
- Add /root/.ssh/id_rsa.pub to /root/.ssh/authorized_keys, make sure this node itself authorized
#### crm cluster join node
- Generate a new rsa key only if /root/.ssh/id_rsa not exist
- Add /root/.ssh/id_rsa.pub to /root/.ssh/authorized_keys, make sure this node itself authorized
- Detect whether login to init node is passwordless, if not, paste /root/.ssh/id_rsa.pub to init node's /root/.ssh/authorized_keys, then **login from join node to init node is passwordless**
- Fetch the public key from init node, and add it to local /root/.ssh/authorized_keys, then **login from init node to join node is passwordless**
#### For multi nodes scenarios
Above solution should works well for two nodes cluster, but not for multi-nodes, image this scenairos:
```
1. A cluster already has node1 and node2
2. node3 join to node1
3. node3 and node1 are passwordless
4. node3 and node2 still not passwordless
```
To address this issue, create function `setup_passwordless_with_other_nodes`, call it at the beginning of `join_cluster`;
In `setup_passwordless_with_other_nodes`, fetch the node list from init node, then swap the key between join node and the other cluster nodes if detect no passwordless configured before.

## Results
- All cluster nodes are passwordless to access each other
- Each node has its own ssh key(existing key or new created rsa key)
- When using --no-overwrite-sshkey option, give warning: `--no-overwrite-sshkey option is deprecated since crmsh does not overwrite ssh keys by default anymore and will be removed in future versions`

### CODE REVIEW

**Comment by zzhou1** [comment]
Well written design! This is much more user friendly to deal with multi nodes.

If we move the scope further to consider the compability to address bsc#1170999#c24, can you describe how to deal with the existant multi-node setup then?

**Comment by liangxin1300** [comment]
> Well written design! This is much more user friendly to deal with multi nodes.
> 
> If we move the scope further to consider the compability to address bsc#1170999#c24, can you describe how to deal with the existant multi-node setup then?

This solution don't have compatibility issue, any newly join node will use its own ssh key, to swap with the init node's public key, then let the init node swap keys with other existing cluster nodes

**Comment by zzhou1** [comment]
> This solution don't have compatibility issue, any newly join node will use its own ssh key, to swap with the init node's public key, then let the init node swap keys with other existing cluster nodes

I wonder when the existent cluster is still old and doesn't have the code to populate the joining node sshkey to the other nodes? Or, the newer joining node will detect crmsh is old on the jumper node and drop-in your stage code `ssh_passwordless_remote` there?

**Comment by liangxin1300** [comment]
> > This solution don't have compatibility issue, any newly join node will use its own ssh key, to swap with the init node's public key, then let the init node swap keys with other existing cluster nodes
> 
> I wonder when the existent cluster is still old and doesn't have the code to populate the joining node sshkey to the other nodes? Or, the newer joining node will detect crmsh is old on the jumper node and drop-in your stage code `ssh_passwordless_remote` there?

Then join process will error and exit with message `No ssh_passwordless_remote stage found on {}, might need to update crmsh on it`

**Comment by zzhou1** [comment]
> > > This solution don't have compatibility issue, any newly join node will use its own ssh key, to swap with the init node's public key, then let the init node swap keys with other existing cluster nodes
> > 
> > 
> > I wonder when the existent cluster is still old and doesn't have the code to populate the joining node sshkey to the other nodes? Or, the newer joining node will detect crmsh is old on the jumper node and drop-in your stage code `ssh_passwordless_remote` there?
> 
> Then join process will error and exit with message `No ssh_passwordless_remote stage found on {}, might need to update crmsh on it`

The error message is informative to guide the user for the next steps. Then, it sounds good to me.

File: crmsh/bootstrap.py [hunk 1]
```diff
@@ -1739,35 +1794,24 @@ def join_ssh(seed_host):
...
-    if got_keys == 0:
-        status("No new SSH keys installed")
-    elif got_keys == 1:
-        status("One new SSH key installed")
-    else:
-        status("%s new SSH keys installed" % (got_keys))
+    configure_local_ssh_key()
+
+    # Detect whether need password to login to seed_host
+    passwordless = False
```
Comment (arbulu89) [inline crmsh/bootstrap.py:1800]: Why do you use this `passwordless` variable? This `if/else` condition looks super strange

Why not simply 
```
if not utils.check_ssh_passwd_need(seed_host):
    status("Configuring SSH keys - This may prompt for root@{}".format(seed_host))
    ...
```

**Comment by liangxin1300** [comment]
> @liangxin1300 To be honest, the code looks extremely hard to understand.
> What I understand is the next. Correct me if I'm wrong
> 
> 1. The `cluster init` process is the same, except now the keys are only created if there are not and `-no-overwrite-sshkey` is deprecated
> 
> And now the difficult part with the join:
> 
> 1. You have created a new stage (`ssh_passwordless_remote`) that basically, loops in the currently existing cluster nodes, and add them the new joining node key, and add the key of the other nodes to the joining node. So here, the difficult part to understand, is that this is executed in a 3rd node. I guess that this is required, as the joining node cannot copy the file to the remote files without asking the password X times.

The `ssh_passwordless_remote` stage, is executed in `init` node. The biggest reason to do this, is avoiding ask the password N-1 times on the newly join node, since init node already setup passwordless with other cluster nodes

> 2. During the join, you check if the connection is possible. If not, create new keys and authorize them.
> 
> The strangest thing is that you call `crm ssh_passwordless_remote` in other node during a crm call. I guess we need to do so, what is super strange.
> 
> Maybe we need to do a live session so you explain me on detail all the changes. What do you think?
> 
> And the issue is what Roger commented. We would have non backward compatibility issues. This looks pretty dangerous. Maybe we need to think how we can share the keys in the other nodes without any new stage, just running `ssh/scp` commands. I wouldn't like to enter in other compatibility issue (more when we are in these situation for that exact thing).
> 
> Putting @diegoakechi in CC. I guess he agrees with my last comment

File: crmsh/bootstrap.py [hunk 2]
```diff
@@ -956,6 +963,14 @@ def append(fromfile, tofile):
             tf.write(ff.read())
 
 
+def append_unique(fromfile, tofile):
+    """
+    Append unique content from fromfile to tofile
+    """
+    if not utils.check_file_content_included(fromfile, tofile):
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:970]: OK, thanks, I will consider that exception

File: crmsh/bootstrap.py [hunk 3]
```diff
@@ -1777,6 +1821,29 @@ def join_ssh(seed_host):
...
+        cmd = "ssh -oStrictHostKeyChecking=no root@{} 'test -f {}'".format(node, public_key_file)
+        if not invoke(cmd):
+            continue
+        _, temp_public_key_file = tmpfiles.create()
+        cmd = "scp -oStrictHostKeyChecking=no root@{}:{} {}".format(node, public_key_file, temp_public_key_file)
+        if not invoke(cmd):
+            error("Failed to run \"{}\"".format(cmd))
+        return temp_public_key_file
+    warn("No ssh key exist on {}".format(node))
+    return None
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1844]: OK

File: crmsh/bootstrap.py [hunk 4]
```diff
@@ -1777,6 +1821,29 @@ def join_ssh(seed_host):
...
+    Fetch public key file from remote node
+    Return a temp file contains public key
+    Return None if no key exist
+    """
+
+    # For dsa, might need to add PubkeyAcceptedKeyTypes=+ssh-dss to config file, see
+    # https://superuser.com/questions/1016989/ssh-dsa-keys-no-longer-work-for-password-less-authentication
+    for key in ("id_rsa", "id_ecdsa", "id_ed25519", "id_dsa"):
+        public_key_file = "/root/.ssh/{}.pub".format(key)
+        cmd = "ssh -oStrictHostKeyChecking=no root@{} 'test -f {}'".format(node, public_key_file)
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1835]: @arbulu89 , This will be called under these scenarios:

- On join node, fetch from init node, just after append join node's public key to init node's authorized_keys, should be passwordless now
- On init node, `ssh_passwordless_remote` stage, fetch from join node, this step was after `join_ssh` which has already setup passwordless between init and join node
- On init node, `ssh_passwordless_remote` stage, fetch from other cluster nodes, assuming that other cluster nodes already configured passwordless between the init node

On above scenarios we all assuming that it's already **should** passwordless configured before running `test -f` remotely.
But yes, you were right, there should also have some `corner case`.

How about here we also use `utils.check_ssh_passwd_need` to detect the connection to this remote node, if just in case there was not passwordless, here we prompted a meaningful message hints we will ask the password multi times.

What do you think about it?

File: crmsh/bootstrap.py [hunk 5]
```diff
@@ -1016,6 +1038,39 @@ def init_ssh_remote():
             append(fn + ".pub", authorized_keys_file)
 
 
+def init_ssh_passwordless_remote():
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1041]: > What is this method for? It's not used anywhere. It's the new stage you defined before?

You can search `globals()["init_" + stage]()` inside `bootstrap_init` function, for now it's the way that crmsh bootstrap to call the stage. Yes, this is where the new stage defined

[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1041]: > In fact, I don't understand what this method is doing, it's quite messy. Which keys are you sharing here?

This method was for swapping the public keys between the current join node and the other cluster nodes. 

> It looks like you add their own public key to their authorized file.

In each loop, we got a `other` cluster member node

1. append join node's public key to the member node's authorized keys;
2. append member node's public key to the join node's authorized keys 

After each loop, the passwordless was setup between the current join node and that cluster member node

> Wouldn't you add the `me` node key?
`me` here means init node, already swapped keys on `join_ssh` method

**Comment by liangxin1300** [comment]
@arbulu89 @diegoakechi @zzhou1 
In a short summary, for `ssh_passwordless_remote` stage:

**pros**: make use of `init` node, to swap public key between join node and the  other cluster existing members, **avoid asking multi times password**(might be N-1 times in worst case)

**cons**: brought new compatibility problem

For now, in this PR, I detected this `compatibility` issue and give an error message:
``` Python
if "Invalid stage" in err:
    error("No ssh_passwordless_remote stage found on {}, might need to update crmsh on it".format(seed_host))
```
To address the compatibility problem here, how about change it like:
``` Python
if "Invalid stage" in err:
    '''
    1. Let init node tell me(the join node) the cluster member list
    2. The join node swap its public key between each cluster member node, here might be asked N-1 times password
    '''
```

File: crmsh/bootstrap.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:970]: I've added these codes in `utils.check_file_content_included`
``` Python
if not os.path.exists(source_file):
    raise ValueError("File {} not exist".format(source_file))
if not os.path.exists(target_file):
    return False
```

File: crmsh/bootstrap.py [hunk 6]
```diff
@@ -1016,6 +1038,39 @@ def init_ssh_remote():
...
+    for node in fetch_nodes_list:
+        public_key_file_remote = fetch_public_key_from_remote_node(node)
+        if not public_key_file_remote:
+            continue
+        # Swap ssh public keys between join node and other cluster nodes
+        append_to_remote_file(public_key_file_join_node, node, AUTHORIZED_KEYS_FILE)
+        append_to_remote_file(public_key_file_remote, join_node, AUTHORIZED_KEYS_FILE)
+
+
+def append_to_remote_file(fromfile, remote_node, tofile):
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1041]: This is something like the format of scp fromfile root@remote_node tofile:)

File: crmsh/bootstrap.py (hunk 1 above)
Comment (liangxin1300) [reply crmsh/bootstrap.py:1776]: Thanks for reminding this, I thought that `passwordless` would be used multi times

Already changed

File: crmsh/bootstrap.py [hunk 7]
```diff
@@ -1739,35 +1794,24 @@ def join_ssh(seed_host):
...
-
-    if got_keys == 0:
-        status("No new SSH keys installed")
-    elif got_keys == 1:
-        status("One new SSH key installed")
-    else:
-        status("%s new SSH keys installed" % (got_keys))
+    configure_local_ssh_key()
+
+    # Detect whether need password to login to seed_host
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Thanks for suggestion! Already create a function named `swap_public_ssh_key`, which can also be reused

File: crmsh/bootstrap.py [hunk 8]
```diff
@@ -987,15 +1002,22 @@ def init_ssh():
...
+    configure_local_ssh_key()
+
+
+def configure_local_ssh_key():
+    """
+    Configure ssh rsa key locally
+
+    If /root/.ssh/id_rsa not exist, generate a new one
+    Add /root/.ssh/id_rsa.pub to /root/.ssh/authorized_keys anyway, make sure itself authorized
+    """
```
Comment (arbulu89) [inline crmsh/bootstrap.py:1014]: I guess that we don't need to check if `/root/.ssh` exists or not, right? `ssh-keygen` should create the folder if needed

File: crmsh/bootstrap.py [hunk 9]
```diff
@@ -1777,6 +1781,52 @@ def join_ssh(seed_host):
...
+        # This time login to remote_node is passwordless
+        append_to_remote_file(RSA_PUBLIC_KEY, remote_node, AUTHORIZED_KEYS_FILE)
+
+    try:
+        # Fetch public key file from remote_node
+        public_key_file_remote = fetch_public_key_from_remote_node(remote_node)
+    except ValueError as err:
+        warn(err)
+        return
+    else:
```
Comment (arbulu89) [inline crmsh/bootstrap.py:1801]: Is here some particular reason why you are using this `else` statement? It looks pretty strange.
The unique reason why I would add it is that `append_unique` can raise a `ValueError` and you don't want to catch it for some reason. Otherwise you could just put it in the `try` block.
Just curiosity hehe

File: crmsh/bootstrap.py [hunk 10]
```diff
@@ -1777,6 +1781,52 @@ def join_ssh(seed_host):
...
 
+def swap_public_ssh_key(remote_node):
+    """
+    Swap public ssh key between remote_node and local
+    """
+    # Detect whether need password to login to remote_node
+    if utils.check_ssh_passwd_need(remote_node):
+        # If no passwordless configured, paste /root/.ssh/id_rsa.pub to remote_node's /root/.ssh/authorized_keys
+        status("Configuring SSH passwordless with root@{}".format(remote_node))
+        # This time login to remote_node is passwordless
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1792]: Yes, I've changed the comment as `# After this, login to remote_node is passwordless`

File: crmsh/bootstrap.py [hunk 11]
```diff
@@ -1777,6 +1781,52 @@ def join_ssh(seed_host):
...
+        warn(err)
+        return
+    else:
+        # Append public key file from remote_node to local's /root/.ssh/authorized_keys
+        # This time login from remote_node is passwordless
+        # Should do this step even passwordless is True, to make sure we got two-way passwordless
+        append_unique(public_key_file_remote, AUTHORIZED_KEYS_FILE)
+
+
+def fetch_public_key_from_remote_node(node):
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py:1808]: Yes

File: crmsh/bootstrap.py (hunk 8 above)
Comment (liangxin1300) [reply crmsh/bootstrap.py:1014]: Yes

File: crmsh/bootstrap.py (hunk 9 above)
Comment (liangxin1300) [reply crmsh/bootstrap.py]: @arbulu89 I've removed the `else` keyword and move forward the codes under `else`
